feature#41 login 로그인 세션 관리 및 구조 수정

soochan8 opened the pull request:
### **📌 Issue**
- #41 

### **🔖 요약**
1. 로그인 세션 관리 (core:auth)
2. ViewModel 책임 분배
3. 구조 수정

### **🛠 작업 내용**
1. 로그인 세션 관리
 - core:Auth 모듈 추가
 - `AuthModule` 생성
 - `AuthRepository, AuthRepositoryImpl`를 통해 로그인 시 세션 관리 로직 추가
 - Flow로 세션 관리
2. ViewModel 책임 분배
 - 기존 `LoginViewModel` 에서 로그인 및 회원 가입 시, 아이디 검사를 진행
 - 변경 후, 로그인 및 회원 가입 호출 시 Repository로 아이디 검사 위임
3. `LoginNavGraph` 수정
 - 기존 LoginScreen에서 viewModel과 state 관리
 - 변경 후, `LoginNavGraph`에서 viewModel과 state 관리 및 LoginScreen으로 위임

### **❗️ 참고사항**
[0b22963](https://github.com/f-lab-edu/commerce_clone/pull/51/commits/0b22963c0fe306020bdb364dca10f3e017e0c794) 
해당 내용 이후는 develop merge 후 발생한 내용입니다!

soochan8 committed:
feat: category/product json 새로 추가

기존 product_list.json 확장성이 힘든 구조로 인하여 category와 product새로 구분

soochan8 committed:
feat: ProductDao, CategoryDao 추가

soochan8 committed:
feat: ProductEntity, CategoryEntity 추가

soochan8 committed:
feat: Product, Category DB 추가

soochan8 committed:
feat: core모듈 공통 ProductVO와 ProductModel 생성

soochan8 committed:
feat: CategoryRepository/Impl 수정 및 remoteDataSource 추가

기존 productDao에서 getAll()로 가져와서 category 필터링
수정 후, categoryDao에서 가져오기

soochan8 committed:
feat: CategoryDetailRepository/Impl 수정

기존 productDao에서 getAll()로 가져와서 category 필터링
수정 후, categoryDao에서 파라미터를 통해 가져오기

soochan8 committed:
delete: 기존 사용하던 카테고리 remoteDataSource 제거

soochan8 committed:
feat: 카테고리 VO, Model, Mapper 수정

soochan8 committed:
feat: categoryDetailMapper 수정

soochan8 committed:
feat: CategoryContract/ViewModel 수정 및 CategoryScreen Model 구조에 맞춰 수정

1. 수정된 categoryDao에서 카테고리 가져오기

soochan8 committed:
feat: CategoryContract ViewModel, Screen 수정

1. 수정된 category, productEntity 구조에 맞춰 변경

soochan8 committed:
feat: category, categoryDetail UIMapper 추가

soochan8 committed:
chore: 마이그레이션 파일 및 build 수정

soochan8 committed:
chore: add coil dependency

soochan8 committed:
feat: 공통 디자인 시스템 추가 (Dimen, Type)

soochan8 committed:
feat: CategoryModel imageUrl 추가

상위 카테고리 imageUrl

soochan8 committed:
feat: 공통 디자인 시스템 및 상위 카테고리 image 설정

soochan8 committed:
feat: CategoryScreen composable 세분화

soochan8 committed:
feat: category 왼쪽(대분류) 스크롤 리스트 composable

soochan8 committed:
feat: category 우측 스크롤 composable

soochan8 committed:
feat: Contract/ViewModel navigation Effect 추가

soochan8 committed:
feat: selectedCategoryId 상태 갱신 시, CategorySidebar Effect 발행

- ViewModel에서 상태 갱신 후, index 계산
- ScrollToSidebar Effcet 발행 및 UI 책임 분리

soochan8 committed:
feat: CategoryScreen 스크롤 및 Effect 처리 추가

- SidebarClick, ContentScroll 책임 분리
- Content Scroll 시, CategorySidebar index로직 ViewModel로 책임 분리

soochan8 committed:
refactor : CategoryDetail NavGraph와 Screen 구조 개선

- 초기 이벤트 및 네비게이션 책임 NavGraph로 이동
- Navigation Effect 추가
- Tab 클릭 시, selectedCategoryTabIndex 매핑 viewModel로 책임 분리

soochan8 committed:
feat: core:auth 모듈 생성

soochan8 committed:
feat: core:auth 모듈 생성

soochan8 committed:
chore: build.gradle 수정 및 settings.gradle auth 모듈 추가

soochan8 committed:
chore: build.gradle 수정 및 settings.gradle auth 모듈 추가

soochan8 committed:
feat: core:auth 모듈에 로그인 세션 관리 기능 추가 및 DI 설정

- 로그인/로그아웃 시 관찰 가능한 sessionFlow 노출
- EncryptedSharedPreferences 기반 구현

soochan8 committed:
feat: core:auth 모듈에 로그인 세션 관리 기능 추가 및 DI 설정

- 로그인/로그아웃 시 관찰 가능한 sessionFlow 노출
- EncryptedSharedPreferences 기반 구현

soochan8 committed:
feat: showError Effect 파라미터 String 수정

soochan8 committed:
feat: showError Effect 파라미터 String 수정

soochan8 committed:
feat: Effect 관련 LaunchedEffect 추가

soochan8 committed:
feat: Effect 관련 LaunchedEffect 추가

soochan8 committed:
feat: 기존 아이디 찾기 제거 및 로그인 Repository/Impl 추가

soochan8 committed:
feat: 기존 아이디 찾기 제거 및 로그인 Repository/Impl 추가

soochan8 committed:
refactor : LoginViewModel 인증 로직 Repository 위임

- 기존 LoginViewModel에서 아이디 조건 검사 진행
- Repository에서 아이디 조건 검사 및 로그인/회원가입 진행
- viewModel은 성공/실패 결과만 받아 처리

soochan8 committed:
refactor : LoginViewModel 인증 로직 Repository 위임

- 기존 LoginViewModel에서 아이디 조건 검사 진행
- Repository에서 아이디 조건 검사 및 로그인/회원가입 진행
- viewModel은 성공/실패 결과만 받아 처리

soochan8 committed:
feat : 앱 재실행 시, 세션 복원 기능 추가

- 세션이 남아있을 경우, 앱 종료 후에도 로그인 상태 유지

soochan8 committed:
feat : 앱 재실행 시, 세션 복원 기능 추가

- 세션이 남아있을 경우, 앱 종료 후에도 로그인 상태 유지

soochan8 committed:
feat: MyPage 모듈 생성

soochan8 committed:
feat: MyPage 모듈 생성

soochan8 committed:
chore: myPage dependency 추가

soochan8 committed:
chore: myPage dependency 추가

soochan8 committed:
feat: Login, MyPage 경로 수정

soochan8 committed:
feat: Login, MyPage 경로 수정

soochan8 committed:
feat: KakaoLogin 성공 시, 파라미터 userId 추가

- 세션 유지를 위해 userId 추가

soochan8 committed:
feat: KakaoLogin 성공 시, 파라미터 userId 추가

- 세션 유지를 위해 userId 추가

soochan8 committed:
feat: LoginContract/ViewModel 세션 확인 State 및 검사 로직 추가

soochan8 committed:
feat: LoginContract/ViewModel 세션 확인 State 및 검사 로직 추가

soochan8 committed:
feat: LoginDestination route 수정

soochan8 committed:
feat: LoginDestination route 수정

soochan8 committed:
feat: LoginNavGraph 세션 여부에 따른 화면 이동

- 세션이 없을 경우 LoginScreen()에서 로그인 진행
- 세션이 있을 경우(로그인 이후) MyPage로 navigate

soochan8 committed:
feat: LoginNavGraph 세션 여부에 따른 화면 이동

- 세션이 없을 경우 LoginScreen()에서 로그인 진행
- 세션이 있을 경우(로그인 이후) MyPage로 navigate

soochan8 committed:
no message

soochan8 committed:
no message

soochan8 committed:
chore: security dependency 추가

soochan8 committed:
chore: security dependency 추가

soochan8 committed:
feat: loginScreen에 state, event 위임

soochan8 committed:
feat: loginScreen에 state, event 위임

soochan8 committed:
feat: LoginScreen에서 위임받은 state와 event로 처리

soochan8 committed:
feat: LoginScreen에서 위임받은 state와 event로 처리

soochan8 committed:
Merge pull request #50 from f-lab-edu/feature/refactor#49_entity개선

기존 ProductEntity 구조 개선 (category/product 세분화) 및 Category 모듈 선적용

soochan8 committed:
Merge remote-tracking branch 'origin/feature/feature#41_login' into feature/feature#41_login

f-lab-pepe commented on feature/login/src/main/java/com/chan/login/data/repository/KakaoLoginManagerImpl.kt:
```diff
@@ -30,7 +30,13 @@ class KakaoLoginManagerImpl @Inject constructor(
+                token != null -> {
+                    UserApiClient.instance.me { user, meError ->
+                        if (user != null) {
+                            trySend(KakaoLoginResult.Success(user.id.toString(), token.accessToken))
```
`UserApiClient.instance.me` 가 비동기인데 밑에 close() 를 바로 부르면 trySend가 실패하지 않나요?

soochan8 replied:
람다를 통해 결과 및 close() 호출을 같이 관리하는 방향으로 개선하겠습니다!

f-lab-pepe commented on feature/login/src/main/java/com/chan/login/data/repository/LoginRepositoryImpl.kt:
```diff
@@ -2,21 +2,53 @@ package com.chan.login.data.repository
 
-    override suspend fun findUserByUserId(userId: String): UserVO? {
-        val user = userDao.findByUsername(userId)
+    override suspend fun appLogin(userId: String, userPw: String): Result<Unit> {
```
storedUser를 반환하고 있는데 Result<Unit> 이 맞나요?

soochan8 replied:
기존 Result로 작업을 했었는데 바꾸지 못한 것 같습니다! 로그인 된 User 정보를 반환해야 하는 것이 맞습니다! 해당 부분도 수정하여 반영하겠습니다!

soochan8 committed:
refactor: KakaoLoginManagerImpl callbackFlow 개선

- 기존 비동기 결과 받기 전에 close()를 호출해서 실패 가능성 존재
- 결과 및 close() 호출 단일 지점에서 관리하여 안정성 향상

soochan8 committed:
feat: errorMessage null safe

soochan8 committed:
feat : appLogin 반환 값 변경 및 ViewModel 처리 로직 수정

- 기존 Result<Unit> > UserVO
- appLogin 후, 로그인 한 계정 id를 바탕으로 session 저장

soochan8 committed:
feat: 문자열 리소스 추가

soochan8 committed:
feat: ErrorMsg 파라미터 StringRes 수정 및 취소 공통 로직 생성

soochan8 commented:
### **🛠 추가 작업 내용**
1. `KakaoLoginManagerImpl` 로직 개선
- 단일 지점 호출로 안정성 및 순서 보장
2. appLogin 반환 값 수정
- 기존 Result<T>에서 T (UserVO)로 수정
- 반환 받은 User정보를 통해 세션 저장

f-lab-pepe commented on feature/login/src/main/java/com/chan/login/navigation/LoginNavGraph.kt:
```diff
@@ -13,7 +22,44 @@ class LoginNavGraph @Inject constructor() : NavGraphProvider {
+                }
+
+                is LoginContract.Effect.ShowError -> {
+                    Log.e("LoginErrorInfo", effect.errorMsg)
```
여기 빌드 되나요?

soochan8 replied:
@f-lab-pepe 
앗..! 아래와 같은 코드로 수정하였습니다!

```
val context = LocalContext.current
val errorMsg = context.getString(effect.errorMsg)
```

해당 부분 제외하고 문제 없다면 merge 진행해도 괜찮을까요?!
